feat: support parameters when creating and executing algorithms

Painguin opened the pull request:
Adds a parameters section for:
- `AddAlgorithm.js`, `EditAlgorithm.js`: the user can add, edit and remove parameters. As he changes the parameters, the python code is modified in real-time to support them.
- `AddExecutionForm.js`: the user chooses the values of the parameters.

The offered parameter types are `Integer`, `Number`, `Float` and `Field Selector` (hierarchical checkboxes to select the fields on which to operate the algorithm, particularly useful for selecting the fields we want to suppress in a dataset).

The suppressing algorithms have been replaced by a single one with a `Field Selector` parameter. The geolocation 2-anonymization has been replaced by a k-anonymization variant where the user can select the parameter `k`.

(make sure to delete the algorithms folder before testing this)

pyphilia commented:
Looks really cool! Quick comments while testing:
In the parameter form:
- It would be better for the default value of parameter name to be as placeholder. 
- It would also be better overall for the form to display errors (eg cannot save two parameters with the same name, cannot save a parameter without name, etc). 
- I think there are a lot of characters to avoid as parameter name. Maybe check for some of them to not exist (the space for instance)
- Not sure the difference between `Integer` and `Number` is really clear. Maybe have `Float` and `Double` instead of `Number`

Painguin commented:
Thank you for your relevant suggestions! I made the changes and it should now work with python 2 :slightly_smiling_face:

hagoptaminian reviewed (commented):
hey @Painguin this is really great. love how much cleaner it makes the algorithms page, and the parameters work very well. also happy to see you're taking up functional components :) left a few comments mainly on UI - please take a look and I'll do another review afterwards. a few other things:

- Make sure you edit the algo descriptions in the algorithms tab. e.g. k-anonymize still says _ensure... there are at least two..._. 'Suppress fields' could become: 'Remove selected fields from the dataset completely.'

- What do you think of making the hash algo also parameterized by field?

- The issue of the former instance of `db.json` not clearing to allow newer versions to be loaded keeps recurring. and it's going to become more cumbersome to reset when we merge @pyphilia's PR which stops us from deleting graasp algos (in which case we have to go into db.json and manually delete the algos). doesn't have to be for this PR, but can you look into somehow resetting/clearing the DB--maybe specifically the algorithms--before loading in the algorithms to prevent this issue from occurring? i.e. to somehow ensure that the 'latest version' of the db is always loaded

hagoptaminian commented on src/components/execution/AddExecutionForm.js:
```diff
@@ -265,7 +281,12 @@ class AddExecutionForm extends Component {
         </FormControl>
-
+        {parameters && parameters.length > 0 && (
+          <SelectParameters
```
(to be clear, I mean that once you select a parameterized algo, the modal pops up, and is then re-accessible via edit parameters)

pyphilia replied:
@hagoptaminian Additionally what JC suggested is to have similar creation/addition of resource. Currently we have 3 methods: modal, new page and in-page form. I would suggest to go for a new page each time we want to create a new entity. Therefore, (and since the execution table is merged) we could have this execution form in a separate page and have the parameters in-page as well. Maybe we can have something similar to the "Add Algorithm" page. What do you think @Painguin @hagoptaminian ?

pyphilia commented on src/components/parameter/FieldSelectorModal.js:
```diff
@@ -0,0 +1,76 @@
+
+  return (
+    <Dialog open={open} onClose={handleClose}>
+      <DialogTitle>{name}</DialogTitle>
```
I agree with Hagop to perhaps add description for parameters in general. It won't be used for python, but rather for the interface.

However I would rather go for the title `Fields Selection {param name}` or `Keys Selection for {param name}` and below the description of the parameter. What do you think @hagoptaminian  ?

pyphilia commented on src/utils/parameter.js:
```diff
@@ -0,0 +1,53 @@
+import { PARAMETER_TYPES } from '../shared/constants';
+
+// verify if the parameter value satisfies the type
+export const isParameterValid = ({ type, value }) => {
```
How do you deal with an undefined type or a type which is not in this list?

pyphilia reviewed (changes requested):
It works really nicely! Some points:

- Do you think it would be possible to display the file once loaded in order to display the parameters as done for "writing from template"?

Painguin commented on src/components/parameter/FieldSelectorModal.js:
```diff
@@ -0,0 +1,76 @@
+import DialogTitle from '@material-ui/core/DialogTitle';
+import { withStyles } from '@material-ui/core/styles';
+import PropTypes from 'prop-types';
+import React from 'react';
```
I agree that the react import should be place first!
The reason it got moved below is because I used an import sorting feature on vscode and I liked how the imports were all sorted afterwards. I don't know if you've already discussed about this but what do you think of enforcing the import order style with eslint @hagoptaminian @pyphilia ? https://dev.to/otamnitram/sorting-your-imports-correctly-in-react-213m

Painguin commented on src/components/execution/AddExecutionForm.js:
```diff
@@ -265,7 +281,12 @@ class AddExecutionForm extends Component {
           />
         </FormControl>
-
+        {parameters && parameters.length > 0 && (
```
Apparently writing `parameters.length` and writing `parameters.length > 0` is not exactly the same. A `0` appears with the first variant:
![image](https://user-images.githubusercontent.com/25949978/101188890-9cc25180-3656-11eb-9f6e-1559be19ca27.png)

pyphilia commented on public/app/listeners/executePythonAlgorithm.js:
```diff
@@ -52,7 +53,30 @@ const executePythonAlgorithm = (mainWindow, db) => (
+    // for each parameter, prepares a pair [--parameter_name, parameter_value] for the command line
+    const preparedParameters =
+      parameters
+        ?.map(({ name, type, value }) => {
```
Then why not have `--k "value"` as one string and not have them added afterwards? I don't know how you deal with `"` after this.

Painguin commented on src/components/parameter/EditParameters.js:
```diff
@@ -0,0 +1,241 @@
+          const invalidName = name.length > 0 && !isParameterNameValid(param);
+          return (
+            // eslint-disable-next-line react/no-array-index-key
+            <Grid container spacing={1} key={paramIdx}>
```
With the name as key, not only the console complains, but very annoying conflicts occur when trying to modify the name (like modifying another textfield instead of the current one, not being focused on the field anymore after entering one character).
I went with both the name and the idx

Painguin replied:
Actually eslint still complains if I try to use the index in the key. I'll keep using only index then and keep the eslint-disable (unless you see another solution :eyes: )

Painguin commented on src/components/algorithm/AddAlgorithm.js:
```diff
@@ -90,23 +91,44 @@ class AddAlgorithm extends Component {
 
+  handleParamsOnChange = (parameters) => {
+    const { code } = this.state;
+    this.setState({ parameters, code: setParametersInCode(code, parameters) });
```
Good idea

Painguin commented on src/components/common/Editor.js:
```diff
@@ -1,11 +1,11 @@
-import PropTypes from 'prop-types';
 import { withTranslation } from 'react-i18next';
 import { EDITOR_PROGRAMMING_LANGUAGES } from '../../config/constants';
+import 'ace-builds/src-noconflict/ext-language_tools';
```
What makes you say that here?

Painguin commented:
Here goes handleParameters v2!
Thank you for your feedback! I've modified the parameter views according to your suggestions:
- I've replaced `hash_users` by `hash_fields` with a field selection parameter.
- The field selector that is shown now depends on the schema. Handling multiple schemas should be more or less supported for the parameters, for when we will provide support for adding schemas.
- I did not move the Add execution into a new page/modal as @pyphilia proposed, but I guess this would not be hard to change.

Let me know what you think of this version!

pyphilia reviewed (changes requested):
Nice work! I let you some new comments.

- When choosing the parameter's value (when executing) could we have buttons in the modal? It is not clear when we close that it was saved or not. That way it would reset when we close or save the changes when we click the save button.
- Could you write some tests for this feature?
- It might be worth adding some comments in the code to let the user know that he shouldn't update the parameters via the code. Do you think it is possible to lock lines in this editor?

pyphilia reviewed (commented):
So much files aha 👍 

- What do you think of refactoring the schemas as a map? That way it would avoid to always `find` by id. 
- Instead of `PropTypes.any.isRequired` you can use `PropTypes.oneOf([...]).isRequired`

hagoptaminian commented:
@Painguin this is working really nicely. I tested the parameterized algorithms several times and they are effective. some suggestions:

- for the 'Edit Parameters' button, consider: (1) choosing a different color, (2) reviewing its alignment/padding

![Screen Shot 2021-01-11 at 9 51 05 AM](https://user-images.githubusercontent.com/19311953/104161375-10444500-53f3-11eb-8df6-467ff1748801.png)

- in the modal: (1) any reason we are still displaying 'hashed_fields'? It looks quite out of place. (2) Add a colon to the end of the descriptions, e.g. _Select fields to hash:_ (3) Why is there a dropdown with the schema? (4) what logic are you using for the default selected fields?

![Screen Shot 2021-01-11 at 10 06 18 AM](https://user-images.githubusercontent.com/19311953/104162435-e2600000-53f4-11eb-9ba7-5faa4f5a56b3.png)

- in the k-anonymization modal, consider using a more compact display for the number input (does it need to stretch across the modal for a number?):

![Screen Shot 2021-01-11 at 10 13 55 AM](https://user-images.githubusercontent.com/19311953/104162928-cb6ddd80-53f5-11eb-8f0e-d6fc55959da9.png)

hagoptaminian commented:
Please edit the algorithm descriptions as follows:

- Hash users: _Hash every occurrence of the 'userId' field in 'actions', 'appInstanceResources', and 'users', and remove all other identifying information from the 'users' key_

- Hash fields: _Perform a SHA-256 hash on selected fields from a dataset_

- Suppress fields: _Completely remove selected fields from a dataset_

- Sanitize users: _Scan an entire dataset for occurrences of user names and user ids, and replace such occurrences with a SHA-256 hash of the corresponding user's userId_

- k-Anonymize geolocation: _Ensure that for every combination of 'country', 'region', and 'city', there are at least k users containing that combination_

- I'm thinking it might be useful to add descriptive Chips to this algorithms page as well. **Hash users** and **Sanitize users** especially are so specific to Graasp datasets that they could use something like 'Optimized for Graasp datasets':
![Screen Shot 2021-01-11 at 10 38 11 AM](https://user-images.githubusercontent.com/19311953/104165076-2a812180-53f9-11eb-84fb-51de37829d1c.png)

hagoptaminian reviewed (commented):
this is working well. left some comments in main conversation thread

pyphilia commented:
- `hashed_fields` is the name of the parameter (so is `k`). I guess `Select fields to hash` is the description, maybe it with more description it might not look that much out of the place.
- I agree with the chip in algorithms, but maybe in another PR, since it will request to save the schema of the ideal datasets in their description. Unless @Painguin think it can go in this already huge PR.

Painguin commented:
Thank you for your reviews!

- I agree that hiding the parameter name for the proposed algorithms is better as the description is enough by itself. However, it makes less sense to adopt that strategy for the user defined ones because it isn't clear that the parameter name will not be shown and it is actually the description that will act as the only hinting piece for referring the parameter (For example, one might think of writing the parameter name without the description thinking he would dissociate multiple parameters with the names and discovers later that only the (empty) description is shown and can't figure out which parameter is which). So I replaced the parameter description field by a "Helper text" field and it is shown instead of the parameter name only when present. 
- The dropdown with the schema is there to support the use of multiple schemas for later. I changed it so that it is only shown if there is more than one schema.
- The default selected fields are distinct for each algorithm. For the graasp algorithms, I have set (in `graaspAlgorithms.js`) the fields that I thought were relevant. For the user algorithms, the default selections corresponds to the one the user set when editing/adding the algorithm.
- I like the idea of the chip but I also think it might be better to put it in another PR. We can think of allowing the addition of chips when adding/editing an algorithm and maybe also the same kind of thing for the datasets.

pyphilia reviewed (approved):
Here some last minor comments ;)

hagoptaminian reviewed (approved):
nice one @Painguin 👍

Painguin commented:
close #24, close #113
